marlowe-spec-test/src/Marlowe/Spec/Core/Arbitrary.hs

```diff
@@ -365,16 +451,229 @@ genTransactionOutput :: InterpretJsonRequest -> GenT IO TransactionOutput
+    , (wLet   , Let <$> liftSemiArbitrary context <*> genValue interpret <*> arbitraryContractWeighted w context interpret)
+    , (wAssert, Assert <$> genObservation interpret <*> arbitraryContractWeighted w context interpret)
+    ]
+  where
+    liftSemiArbitrary :: SemiArbitrary a => Context -> GenT IO a
+    liftSemiArbitrary = liftGen . semiArbitrary
+arbitraryContractWeighted [] _ _ = pure Close
+
+-- | Default weights for contract terms.
+defaultContractWeights :: (Int, Int, Int, Int, Int, Int)
+defaultContractWeights = (35, 20, 10, 15, 20, 5)
+
+-- | Generate a semi-random contract of a given depth.
+arbitraryContractSized :: Int           -- ^ The maximum depth.
+                       -> Context       -- ^ The Marlowe context.
+                       -> InterpretJsonRequest
+                       -> GenT IO Contract  -- ^ Generator for a contract.
+arbitraryContractSized = arbitraryContractWeighted . (`replicate` defaultContractWeights)
+
+-- | Some value identifiers.
+randomValueIds :: [ValueId]
+randomValueIds =
+  [
+    ValueId "x"
+  , ValueId "id"
+  , ValueId "lab"
+  , ValueId "idea"
+  , ValueId "story"
+  , ValueId "memory"
+  , ValueId "fishing"
+  , ValueId ""
+  , ValueId "drawing"
+  , ValueId "reaction"
+  , ValueId "difference"
+  , ValueId "replacement"
+  , ValueId "paper apartment"
+  , ValueId "leadership information"
+  , ValueId "entertainment region assumptions"
+  , ValueId "candidate apartment reaction replacement"  -- NB: Too long for ledger.
+  ]
+
+-- | Class for arbitrary values with respect to a context.
+class Arbitrary a => SemiArbitrary a where
+  -- | Generate an arbitrary value within a context.
+  semiArbitrary :: Context -> Gen a
+  semiArbitrary context =
+     case fromContext context of
+       [] -> arbitrary
+       xs -> perturb arbitrary xs
+  -- | Report values present in a context.
+  fromContext :: Context -> [a]
+  fromContext _ = []
+
+-- | Select an element of a list with high probability, or create a non-element at random with low probability.
+perturb :: Gen a   -- ^ The generator for a random item.
+        -> [a]     -- ^ The list of pre-defined items.
+        -> Gen a   -- ^ Generator for an item
+perturb gen [] = gen
+perturb gen xs = frequency [(20, gen), (80, elements xs)]
+
+-- | Context for generating correlated Marlowe terms and state.
+data Context =
+  Context
+  {
+    amounts      :: [Arith.Int]                  -- ^ Universe of token amounts.
+  , choiceNames  :: [String]                     -- ^ Universe of choice names.
+  , chosenNums   :: [Arith.Int]                  -- ^ Universe of chosen numbers.
+  , valueIds     :: [ValueId]                    -- ^ Universe of value identifiers.
+  , values       :: [Arith.Int]                  -- ^ Universe of values.
+  , times        :: [Arith.Int]                  -- ^ Universe of times.
+  , cboundValues :: Map ValueId Arith.Int        -- ^ Bound values for state.
+  }
+
+instance Prelude.Ord ValueId where
```

Comment:
There is a `CoreOrphanEq` module in the `marlowe/isabelle/haskell` folder. We could create a similar file for `Ord`, rename the file to `CoreOphanInstances` or use the trick explained [here](https://isabelle.zulipchat.com/#narrow/stream/238552-Beginner-Questions/topic/Force.20Eq.20instance.20export/near/306701955). But I would avoid having an orphan instance in this file.